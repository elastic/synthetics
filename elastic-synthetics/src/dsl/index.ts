import { Journey, JourneyCallback, JourneyOptions } from './journey'
import { Step, StepCallback } from './step'

class State {
    journeys: Journey[]
    currentJourney?: Journey

    constructor() {
        this.reset();
    }

    public reset() {
        this.journeys = [];
        this.currentJourney = undefined;
    }
}

export const state = new State();

export const journey = (options: JourneyOptions | string, callback: JourneyCallback) => {
    if (typeof options === 'string') {
        options = {name: options, id: options}
    }
    const j = new Journey(options, callback);
    state.journeys.push(j)
    state.currentJourney = j;

    j.callback(); // load steps
}

export const step = (name: string, callback: StepCallback) => {
    const step = new Step(name, callback);
    state.currentJourney.steps.push(step)
}