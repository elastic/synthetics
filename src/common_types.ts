import { Protocol } from 'playwright/types/protocol';

export type StatusValue = 'succeeded' | 'failed' | 'skipped';

export type FilmStrip = {
  snapshot: string;
  name: string;
  ts: number;
};

export type NetworkInfo = {
  url: string;
  method: string;
  type: string;
  request: Protocol.Network.Request;
  response: Protocol.Network.Response;
  isNavigationRequest: boolean;
  start: number;
  end: number;
  status: number;
};
