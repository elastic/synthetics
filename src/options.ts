/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import merge from 'deepmerge';
import { readConfig } from './config';
import { parseNetworkConditions } from './helpers';
import type { CliArgs, RunOptions } from './common_types';

/**
 * Set debug based on DEBUG ENV and -d flags
 * namespace - synthetics
 */
const namespace = 'synthetics';
if (process.env.DEBUG && process.env.DEBUG.includes(namespace)) {
  process.env.DEBUG = '1';
}

export function normalizeOptions(cliArgs: CliArgs): RunOptions {
  const options: RunOptions = {
    ...cliArgs,
    environment: process.env['NODE_ENV'] || 'development',
  };
  /**
   * Group all events that can be consumed by heartbeat and
   * eventually by the Synthetics UI.
   */
  if (cliArgs.richEvents) {
    options.reporter = cliArgs.reporter ?? 'json';
    options.ssblocks = true;
    options.network = true;
    options.trace = true;
    options.quietExitCode = true;
  }

  if (cliArgs.capability) {
    const supportedCapabilities = [
      'trace',
      'network',
      'filmstrips',
      'metrics',
      'ssblocks',
    ];
    /**
     * trace - record chrome trace events(LCP, FCP, CLS, etc.) for all journeys
     * network - capture network information for all journeys
     * filmstrips - record detailed filmstrips for all journeys
     * metrics - capture performance metrics (DOM Nodes, Heap size, etc.) for each step
     * ssblocks - Dedupes the screenshots in to blocks to save storage space
     */
    for (const flag of cliArgs.capability) {
      if (supportedCapabilities.includes(flag)) {
        options[flag] = true;
      } else {
        console.warn(
          `Missing capability "${flag}", current supported capabilities are ${supportedCapabilities.join(
            ', '
          )}`
        );
      }
    }
  }

  /**
   * Validate and read synthetics config file
   * based on the environment
   */
  const config =
    cliArgs.config || !cliArgs.inline
      ? readConfig(options.environment, cliArgs.config)
      : {};

  options.params = Object.freeze(merge(config.params, cliArgs.params || {}));

  /**
   * Favor playwright options passed via cli to inline playwright options
   */
  options.playwrightOptions = merge.all([
    config.playwrightOptions || {},
    cliArgs.playwrightOptions || {},
    {
      headless: cliArgs.headless,
      chromiumSandbox: cliArgs.sandbox,
      ignoreHTTPSErrors: cliArgs.ignoreHttpsErrors,
    },
  ]);

  if (cliArgs.throttling) {
    options.networkConditions = parseNetworkConditions(
      cliArgs.throttling as string
    );
  }

  return options;
}
