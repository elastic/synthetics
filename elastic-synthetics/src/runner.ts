import { state, journey, step } from './dsl';
import * as playwright from 'playwright';
import { Command } from 'commander';
import { readFileSync } from 'fs';
import { createInterface as createReadlineInterface } from 'readline'


const browserType = 'chromium'

export const runJourneys = async (suiteParams: any) => {
    console.debug("Running with suite params", suiteParams);
    const browser = await playwright[browserType].launch({ headless: false });
    const context = await browser.newContext()
    const page = await context.newPage()
    console.log(`Found ${state.journeys.length} journeys`)
    for (let i = 0; i < state.journeys.length; i++) {
        const journey = state.currentJourney = state.journeys[i]
        console.log(`Journey: ${journey.options.name}`)
        for (let j = 0; j < journey.steps.length; j++) {
            const step = journey.steps[j]
            console.log(`Step: ${step.name}`)
            await step.callback(page, suiteParams, {context, browser})
        }
    }

    await browser.close()
}

export const run = async (aroundHook: (runSteps: (suiteParams: any) => void, environment: string, suiteParams: any) => void = (r) => { r({}) }) => {
    const program = new Command;
    let argSuiteParams = {};
    program
        .command("run [file]")
        .option('-s, --suite-params <jsonstring>')
        .option('-e, --environment <envname>', 'e.g. production', 'development')
        .option('-j, --json', 'output newline delimited JSON')
        .option('--stdin', 'read script file input from stdin')
        .description("Run the tests")
        .action(async (file, options) => {
            if (options.suiteParams) {
                const parsedSuiteParams = JSON.parse(options.suiteParams);
                console.log("SPARAMS", parsedSuiteParams)
                argSuiteParams = {...argSuiteParams, ...parsedSuiteParams};
            }

            const singleMode = options.stdin || file;
            if (singleMode) {
                state.reset()
                const source = options.stdin ? await readStdin() : readFileSync(file, {encoding: 'utf8'});
                console.log("Running single script...", source)
                loadInlineScript(source, argSuiteParams);
            }

            console.log("Running...")

            aroundHook((suiteParams) => runJourneys(suiteParams), options.environment, argSuiteParams)
        })

    program.parse()
}

const readStdin = async (): Promise<string> => {
    let source = "";
    const rl = createReadlineInterface({input: process.stdin})
    rl.on('line', (line) => {
        source += line + "\n";
    });

    return new Promise<string>((resolve) => {
        rl.on('close', () => {
            resolve(source)
        })
    });
}

const loadInlineScript = (source, argSuiteParams) => {
    const scriptFn = new Function('step', 'suiteParams', 'console', source);
    journey('inline', async () => {
        console.log("MAKING STEPS")
        scriptFn.apply(null, [step, argSuiteParams, console])
    })
    console.log("r STEPS")
}