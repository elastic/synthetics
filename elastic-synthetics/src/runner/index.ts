import { state, journey, step } from '../dsl';
import * as playwright from 'playwright';
const browserType = 'chromium'

export const runJourneys = async (suiteParams: any) => {
    console.debug("Running with suite params", suiteParams);
    const browser = await playwright[browserType].launch({ headless: false });
    const context = await browser.newContext()
    const page = await context.newPage()
    console.log(`Found ${state.journeys.length} journeys`)
    for (let i = 0; i < state.journeys.length; i++) {
        const journey = state.currentJourney = state.journeys[i]
        console.log(`Journey: ${journey.options.name}`)
        for (let j = 0; j < journey.steps.length; j++) {
            const step = journey.steps[j]
            console.log(`Step: ${step.name}`)
            await step.callback(page, suiteParams, {context, browser})
        }
    }

    await browser.close()
}
