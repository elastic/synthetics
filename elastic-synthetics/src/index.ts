require('source-map-support').install();

export * from './dsl';
export * from './runner';

import { run } from './runner' ;

if (require.main === module)  {
    run()
}
