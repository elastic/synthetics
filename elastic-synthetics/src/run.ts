#!/usr/bin/env node

require('source-map-support').install();
import { runJourneys } from './runner'
import { Command } from 'commander';
import { readFileSync } from 'fs';
import { journey, step, state } from './dsl';
import { createInterface as createReadlineInterface } from 'readline'

export const run = async (aroundHook: (runSteps: (suiteParams: any) => void, environment: string, suiteParams: any) => void = (r) => { r({}) }) => {
    const program = new Command;
    let argSuiteParams = {};
    program
        .command("run [file]")
        .option('-s, --suite-params <jsonstring>')
        .option('-e, --environment <envname>', 'e.g. production', 'development')
        .option('-j, --json', 'output newline delimited JSON')
        .option('--stdin', 'read script file input from stdin')
        .description("Run the tests")
        .action(async (file, options) => {
            if (options.suiteParams) {
                const parsedSuiteParams = JSON.parse(options.suiteParams);
                argSuiteParams = {...argSuiteParams, ...parsedSuiteParams};
            }

            const singleMode = options.stdin || file;
            if (singleMode) {
                state.reset()
                const source = options.stdin ? await readStdin() : readFileSync(file, {encoding: 'utf8'});
                console.log("Running single script...", source)
                loadInlineScript(source, argSuiteParams);
            }

            console.log("Running...")

            aroundHook((suiteParams) => runJourneys(suiteParams), options.environment, argSuiteParams)
        })

    program.parse()
}

const readStdin = async (): Promise<string> => {
    let source = "";
    const rl = createReadlineInterface({input: process.stdin})
    rl.on('line', (line) => {
        source += line + "\n";
    });

    return new Promise<string>((resolve) => {
        rl.on('close', () => {
            resolve(source)
        })
    });
}

const loadInlineScript = (source, argSuiteParams) => {
    const scriptFn = new Function('step', 'suiteParams', 'console', source);
    journey('inline', async () => {
        console.log("MAKING STEPS")
        scriptFn.apply(null, [step, argSuiteParams, console])
    })
    console.log("r STEPS")
}

run()