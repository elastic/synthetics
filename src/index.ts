/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { runner } from './core';
import { RunOptions } from './core/runner';
import { setLogger } from './core/logger';
import { parseArgs } from './parse_args';

/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install();

export async function run(options: RunOptions) {
  const cliArgs = parseArgs();
  /**
   * Use the NODE_ENV variable to control the environment if its not explicity
   * passed from either CLI or through the API
   */
  options.environment = options.environment || process.env['NODE_ENV'];
  /**
   * set up logger with appropriate file descriptor
   * to capture all the DEBUG logs when running from heartbeat
   */
  const outfd = options.outfd ?? cliArgs.outfd;
  setLogger(outfd);

  try {
    return await runner.run({
      ...options,
      headless: options.headless ?? cliArgs.headless,
      screenshots: options.screenshots ?? cliArgs.screenshots,
      dryRun: options.dryRun ?? cliArgs.dryRun,
      journeyName: options.journeyName ?? cliArgs.journeyName,
      network: options.network ?? cliArgs.network,
      pauseOnError: options.pauseOnError ?? cliArgs.pauseOnError,
      reporter: cliArgs.json && !options.reporter ? 'json' : options.reporter,
      outfd,
    });
  } catch (e) {
    console.error('Failed to run the test', e);
    process.exit(1);
  }
}

export { beforeAll, afterAll, journey, step, before, after } from './core';
