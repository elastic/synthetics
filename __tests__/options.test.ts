/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { CliArgs } from '../src/common_types';
import { normalizeOptions } from '../src/options';
import { join } from 'path';

describe('options', () => {
  it('normalize', async () => {
    const cliArgs: CliArgs = {
      params: {
        foo: 'bar',
      },
      headless: true,
      sandbox: false,
      screenshots: 'on',
      dryRun: true,
      match: 'check*',
      pauseOnError: true,
      config: join(__dirname, 'fixtures', 'synthetics.config.ts'),
    };
    expect(normalizeOptions({})).toMatchObject({
      environment: 'test',
      params: {},
    });
    expect(normalizeOptions(cliArgs)).toMatchObject({
      dryRun: true,
      environment: 'test',
      match: 'check*',
      params: {
        foo: 'bar',
        url: 'non-dev',
      },
      pauseOnError: true,
      playwrightOptions: {
        chromiumSandbox: false,
        defaultBrowserType: 'chromium',
        deviceScaleFactor: 4.5,
        hasTouch: true,
        headless: true,
        ignoreHTTPSErrors: undefined,
        isMobile: true,
        userAgent:
          'Mozilla/5.0 (Linux; Android 8.0.0; SM-G965U Build/R16NW) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4929.0 Mobile Safari/537.36',
        viewport: {
          height: 658,
          width: 320,
        },
      },
      screenshots: 'on',
    });
  });
});
