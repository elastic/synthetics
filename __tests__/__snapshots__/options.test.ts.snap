// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options normalize 1`] = `
Object {
  "config": "/Users/vigneshh/elastic/synthetics/__tests__/fixtures/synthetics.config.ts",
  "dryRun": true,
  "environment": "test",
  "headless": true,
  "match": "check*",
  "params": Object {
    "foo": "bar",
    "url": "non-dev",
  },
  "pauseOnError": true,
  "playwrightOptions": Object {
    "chromiumSandbox": false,
    "defaultBrowserType": "chromium",
    "deviceScaleFactor": 4.5,
    "hasTouch": true,
    "headless": true,
    "ignoreHTTPSErrors": undefined,
    "isMobile": true,
    "userAgent": "Mozilla/5.0 (Linux; Android 8.0.0; SM-G965U Build/R16NW) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4595.0 Mobile Safari/537.36",
    "viewport": Object {
      "height": 658,
      "width": 320,
    },
  },
  "sandbox": false,
  "screenshots": "on",
}
`;
